pyccel/ast/numpyext.py

```diff
@@ -416,13 +423,36 @@ def Shape(arg):
         return PythonTuple(*arg.shape)
 
 #==============================================================================
+
 class NumpyImag(PythonImag):
-    """Represents a call to  numpy.real for code generation.
+    """Represents a call to  numpy.imag for code generation.
 
     > a = 1+2j
     > np.imag(a)
     2.0
     """
+    def __new__(cls, arg):
+        if not isinstance(arg.dtype, NativeComplex):
+            dtype=NativeInteger() if isinstance(arg.dtype, NativeBool) else arg.dtype
+            if isinstance(arg, (Literal, Variable)) and arg.rank == 0:
+                return convert_to_literal(0, dtype, arg.precision)
+            return NumpyZeros(arg.shape, dtype=dtype)
+        return super().__new__(cls, arg)
+
+    def __init__(self, arg):
+        super().__init__(arg)
+        self._precision = arg.precision
+        self._order = arg.order
+        self._dtype = arg.dtype if not isinstance(arg.dtype, NativeComplex) else NativeReal()
```

Comment:
Is this if ternary useful? If the dtype is not `NativeComplex` then `__new__` should return a different type so this `__init__` will never be called. Or have I missed something?

Reply:
what if dtype is `NativeComplex` then return will be a complex type

Reply:
To be clear, my suggested fix is:
```
self._dtype = NativeReal()
```

The question is when is the dtype not `NativeComplex`?

Reply:
When `__new__` returns `super().__new__(...)` then the `__init__` of the class is called. When `__new__` returns anything else then it isn't. So I think `NumpyImag.__init__` should only be called if the dtype of the argument is `NativeComplex()`?

Reply:
yes, you are right.